plugin/src/main/java/io/jenkins/plugins/coverage/metrics/restapi/ModifiedLinesCoverageApi.java

```diff
@@ -0,0 +1,97 @@
+     *         containing the file tree.
+     *
+     * @return a list of {@link FileWithModifiedLines} objects.
+     */
+    private static List<FileWithModifiedLines> createListOfFilesWithModifiedLines(final Node node) {
+        var filesWithModifiedLines = new ArrayList<FileWithModifiedLines>();
+
+        for (FileNode fileNode : node.filterByModifiedLines().getAllFileNodes()) {
+            var missedLines = new ArrayList<>(fileNode.getMissedLines());
+            var partiallyCoveredLines = new ArrayList<>(fileNode.getPartiallyCoveredLines().keySet());
+            var coveredLines = new ArrayList<Integer>();
+
+            for (Integer line : fileNode.getLinesWithCoverage()) {
+                if (fileNode.getMissedOfLine(line) == 0 && fileNode.getCoveredOfLine(line) > 0) {
+                    coveredLines.add(line);
+                }
+            }
+
+            var modifiedLinesBlocks = new TreeSet<ModifiedLinesBlock>();
+
+            calculateModifiedLineBlocks(coveredLines, modifiedLinesBlocks, LineCoverageType.COVERED);
+            calculateModifiedLineBlocks(missedLines, modifiedLinesBlocks, LineCoverageType.MISSED);
+            calculateModifiedLineBlocks(partiallyCoveredLines, modifiedLinesBlocks, LineCoverageType.PARTIALLY_COVERED);
+
+            var changedFile = new FileWithModifiedLines(fileNode.getRelativePath(), modifiedLinesBlocks);
+            filesWithModifiedLines.add(changedFile);
+        }
+
+        return filesWithModifiedLines;
+    }
+
+    /**
+     * This method parses over the modified lines of a file and combines consecutive line numbers with the same coverage
+     * type into a {@link ModifiedLinesBlock} object. Result is sorted by starting line number.
+     *
+     * @param modifiedLines
+     *         list containing the integer numbers of modified lines.
+     * @param modifiedLinesBlocks
+     *         list containing modifiedLinesBlock objects.
+     * @param type
+     *         type of coverage pertaining to each line of code ({@link LineCoverageType#COVERED},
+     *         {@link LineCoverageType#MISSED}, or {@link LineCoverageType#PARTIALLY_COVERED})
+     */
+    private static void calculateModifiedLineBlocks(final List<Integer> modifiedLines,
+            final SortedSet<ModifiedLinesBlock> modifiedLinesBlocks, final LineCoverageType type) {
+
+        if (modifiedLines.isEmpty()) {
+            return;
+        }
+
+        int currentLine = modifiedLines.get(0);
+        for (int i = 0; i < modifiedLines.size(); i++) {
+            if (i == modifiedLines.size() - 1 || !modifiedLines.get(i).equals(modifiedLines.get(i + 1) - 1)) {
+
+                var modifiedLinesBlock = new ModifiedLinesBlock(currentLine, modifiedLines.get(i), type);
+                modifiedLinesBlocks.add(modifiedLinesBlock);
+
+                if (i < modifiedLines.size() - 1) {
+                    currentLine = modifiedLines.get(i + 1);
+                }
+            }
+        }
+    }
```

Comment:
I'm not sure if this is relevant: the result is not accurate as typically not all source code lines have a coverage counter (comments, empty lines). As a user I would assume that those not measured lines would not create a new lines block. See https://github.com/jenkinsci/coverage-model/blob/67df3455cb8577a872eecef5c8daae0564c445a7/src/main/java/edu/hm/hafner/coverage/FileNode.java#L576 for an alternative implementation of this problem.

Reply:
Yes true, I implemented another solution that should fix this problem.

Reply:
I improved the implementation like this:
- the API will provide only coverage information for modified lines (as described, has been wrong before)
- lines without coverage will be included in a block if:
  - the lines with coverage information before and after have the same coverage type
  - the lines without coverage information have also been modified (else, there is too much not required information since the user is only interested in modified lines here, in my opinion)
- I added Unit tests to cover this behavior